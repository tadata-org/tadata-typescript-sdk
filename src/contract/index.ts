export { tadataContract, type TadataContract } from './tadata-contract';
export { createApiClient } from './client';
