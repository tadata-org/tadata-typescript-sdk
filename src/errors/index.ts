/**
 * Base error class for all Tadata SDK errors
 */
export class TadataError extends Error {
  constructor(
    readonly message: string,
    readonly code: string,
    readonly statusCode?: number,
    readonly cause?: Error | unknown
  ) {
    super(message);
    this.name = 'TadataError';

    // Maintain proper stack trace for where error was thrown
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }
}

/**
 * Error thrown when the OpenAPI specification is invalid
 */
export class SpecInvalidError extends TadataError {
  constructor(message: string, details?: unknown, cause?: Error | unknown) {
    super(message, 'spec_invalid', 400, cause);
    this.name = 'SpecInvalidError';
    this.details = details;
  }

  readonly details?: unknown;
}

/**
 * Error thrown when authentication fails
 */
export class AuthError extends TadataError {
  constructor(message = 'Authentication failed', cause?: Error | unknown) {
    super(message, 'auth_error', 401, cause);
    this.name = 'AuthError';
  }
}

/**
 * Error thrown for API errors
 */
export class ApiError extends TadataError {
  constructor(
    message: string,
    statusCode: number,
    readonly body?: unknown,
    cause?: Error | unknown
  ) {
    super(message, 'api_error', statusCode, cause);
    this.name = 'ApiError';
  }
}

/**
 * Error thrown for network-related failures
 */
export class NetworkError extends TadataError {
  constructor(message: string, cause?: Error | unknown) {
    super(message, 'network_error', undefined, cause);
    this.name = 'NetworkError';
  }
}
