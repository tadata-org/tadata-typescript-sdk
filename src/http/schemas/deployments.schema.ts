import { z } from 'zod';
import { createApiResponseSchema } from './response.schema';

export const DeploymentResponseMinSchema = z.object({
  id: z.string(),
  createdAt: z
    .union([z.string(), z.date()])
    .transform(val => (val instanceof Date ? val.toISOString() : val))
    .optional(),
  createdBy: z.string(),
  updatedBy: z.string(),
  mcpServerId: z.string(),
  openAPISpecHash: z.string(),
  mcpSpecHash: z.string(),
  status: z.string(),
});

// Define a Zod schema for OpenAPI 3.0 with basic validation
export const OpenApi3Schema = z
  .object({
    openapi: z.string().startsWith('3.'),
    info: z
      .object({
        title: z.string(),
        version: z.string(),
      })
      .passthrough(),
    paths: z.record(z.string(), z.unknown()),
  })
  .passthrough();

export const UpsertDeploymentBodySchema = z.object({
  openApiSpec: OpenApi3Schema,
  name: z.string().optional(),
  baseUrl: z.string().optional(),
});

// Original unwrapped response schema
const UpsertDeploymentResponseDataSchema = z.object({
  updated: z.boolean(),
  deployment: DeploymentResponseMinSchema,
});

// API Response envelope wrapped schema
export const UpsertDeploymentResponseSchema = createApiResponseSchema(
  UpsertDeploymentResponseDataSchema
);
