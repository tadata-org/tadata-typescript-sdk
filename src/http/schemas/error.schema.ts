import { z } from 'zod';

/**
 * Strictly typed error codes for consistent error handling
 * Matches the implementation in the @tadata/shared package
 */
export enum ErrorCode {
  VALIDATION_ERROR = 'VALIDATION_ERROR',
  SERVICE_VALIDATION_ERROR = 'SERVICE_VALIDATION_ERROR',
  JSON_PARSE_ERROR = 'JSON_PARSE_ERROR',
  INVALID_CONTENT_TYPE = 'INVALID_CONTENT_TYPE',
  AUTH_ERROR = 'AUTH_ERROR',
  NOT_FOUND = 'NOT_FOUND',
  INTERNAL_ERROR = 'INTERNAL_ERROR',
}

// Define ValidationErrorSchema or import if it exists elsewhere - assuming it for now
const ValidationErrorSchema = z.object({
  field: z.string(),
  message: z.string(),
  source: z.string().optional(), // Or more specific if applicable
});

export const ApiErrorSchema = z.object({
  status: z.number(),
  code: z.nativeEnum(ErrorCode), // Use the ErrorCode enum instead of z.string()
  message: z.string(),
  errors: z.array(ValidationErrorSchema).optional(),
  details: z.unknown().optional(),
});
