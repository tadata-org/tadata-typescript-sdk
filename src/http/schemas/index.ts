export * from './error.schema';
export * from './deployments.schema';
export * from './response.schema';
