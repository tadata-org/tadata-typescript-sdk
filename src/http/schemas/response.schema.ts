import { z } from 'zod';
import { ApiErrorSchema, ErrorCode } from './error.schema';

/**
 * Standard API response envelope that works for both success and error cases
 * T is the type of the data payload for successful responses
 */
export const createApiResponseSchema = <T extends z.ZodTypeAny>(dataSchema: T) =>
  z
    .object({
      ok: z.boolean(),
      status: z.number(),
      data: dataSchema.optional(),
      error: ApiErrorSchema.omit({ status: true }).optional(),
    })
    .refine(
      data => (data.ok && data.data !== undefined) || (!data.ok && data.error !== undefined),
      {
        message: "Response must have 'data' when ok is true and 'error' when ok is false",
      }
    );

/**
 * Helper to create a successful response envelope
 */
export function createSuccessResponse<T>(status: number, data: T) {
  return {
    ok: true,
    status,
    data,
  };
}

/**
 * Helper to create an error response envelope
 */
export function createErrorResponse(error: z.infer<typeof ApiErrorSchema>) {
  const response: {
    ok: false;
    status: number;
    error: {
      code: ErrorCode;
      message: string;
      errors?: any[];
      details?: unknown;
    };
  } = {
    ok: false,
    status: error.status,
    error: {
      code: error.code,
      message: error.message,
    },
  };

  // Conditionally add errors and details only if they exist
  if (error.errors) {
    response.error.errors = error.errors;
  }

  if (error.details) {
    response.error.details = error.details;
  }

  return response;
}
