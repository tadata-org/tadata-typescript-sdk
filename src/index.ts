// Export resources
export * from './resources';

// Core exports
export * from './core/sdk';
export * from './core/logger';

// Error exports
export * from './errors';

// OpenAPI exports
export * from './openapi';

// HTTP exports
export * from './http/contracts';
export * from './http/schemas/error.schema';
export * from './http/schemas/response.schema';
