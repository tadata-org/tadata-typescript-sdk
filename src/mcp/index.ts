export { McpResource, type McpDeployInput } from './mcp-resource';
export { type McpDeploymentResult } from './types';
