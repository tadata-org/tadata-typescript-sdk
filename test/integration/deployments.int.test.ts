import { StatusCodes } from 'http-status-codes';
import * as nock from 'nock';
import { v4 as uuidv4 } from 'uuid';
import { OpenApiSource, SpecInvalidError, Tadata } from '../../src';
import { ErrorCode } from '../../src/http/schemas';
import { createErrorResponse, createSuccessResponse } from '../utils/response-helpers';

const validOpenApiSpec = {
  openapi: '3.0.0',
  info: {
    title: 'Test API',
    version: '1.0.0',
  },
  paths: {},
};

const invalidOpenApiSpec = {
  info: {
    title: 'Test API',
    version: '1.0.0',
  },
  paths: {},
};

const TEST_API_KEY = 'test_api_key';
const INVALID_API_KEY = 'invalid_api_key';

const BASE_URL = 'https://api.tadata.com';

// Force nock to use the same URL construction that axios uses
// This ensures that the interceptors match the actual requests
nock.enableNetConnect('127.0.0.1'); // Allow localhost connections for CI/CD if needed

describe('Deployments Integration Test (Nock)', () => {
  const sdk = new Tadata({
    apiKey: TEST_API_KEY,
  });

  const invalidSdk = new Tadata({
    apiKey: INVALID_API_KEY,
  });

  beforeAll(() => {
    // Strictly disable all network connections except those we explicitly mock
    nock.disableNetConnect();
    // Optionally allow localhost for debugging
    nock.enableNetConnect('127.0.0.1');
  });

  afterEach(() => {
    nock.cleanAll();

    // Ensure no pending mocks
    if (!nock.isDone()) {
      console.error('Not all nock interceptors were used!');
      nock.cleanAll();
    }
  });

  afterAll(() => {
    nock.enableNetConnect();
    nock.restore();
  });

  describe('when deploying an OpenAPI spec', () => {
    it('should successfully deploy an MCP server', async () => {
      const mockDeploymentId = uuidv4();
      const mockCreatedAt = new Date().toISOString();
      const serviceNameInMockedResponse = 'Service Name From Mocked Server Response';

      const mockServerResponseData = {
        updated: true,
        deployment: {
          id: mockDeploymentId,
          name: serviceNameInMockedResponse,
          updated: true,
          createdAt: mockCreatedAt,
          createdBy: 'test-user',
          updatedBy: 'test-user',
          mcpServerId: uuidv4(),
          openAPISpecHash: 'hash123',
          mcpSpecHash: 'hash456',
          status: 'active',
        },
      };
      const successResponseEnvelope = createSuccessResponse(
        StatusCodes.CREATED,
        mockServerResponseData
      );

      // Mock the POST request to the deployments endpoint
      const scope = nock(BASE_URL)
        .post('/api/deployments/from-openapi')
        .query(true) // Match any query params
        .reply(function (uri) {
          // Verify apiKey is in the query as expected
          const url = new URL(`${BASE_URL}${uri}`);
          const apiKey = url.searchParams.get('apiKey');

          if (apiKey !== TEST_API_KEY) {
            return [
              401,
              createErrorResponse({
                status: 401,
                code: ErrorCode.AUTH_ERROR,
                message: 'Invalid API key from mock',
              }),
            ];
          }

          return [StatusCodes.CREATED, successResponseEnvelope];
        });

      const source = OpenApiSource.fromObject(validOpenApiSpec);
      const requestedDeploymentName = 'Test API Service from Nock';

      const result = await sdk.mcp.deploy({
        spec: source,
        name: requestedDeploymentName,
        apiBaseUrl: 'https://example.com/api',
      });

      // Validate expected response structure
      expect(result).toBeDefined();
      expect(result.id).toBe(mockDeploymentId);
      expect(result.updated).toBe(true);
      expect(result.createdAt).toBeInstanceOf(Date);
      expect(Date.now() - result.createdAt.getTime()).toBeLessThan(60000);

      const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
      expect(uuidRegex.test(result.id)).toBe(true);

      expect(scope.isDone()).toBe(true);
    });

    it('should throw SpecInvalidError for invalid OpenAPI specs before making API call', () => {
      // This test doesn't need scope since it's testing client-side validation
      expect(() => OpenApiSource.fromObject(invalidOpenApiSpec)).toThrow(SpecInvalidError);
    });

    it('should throw SpecInvalidError for backend validation failures', async () => {
      const originalBackendErrorMessage = 'Invalid OpenAPI specification from backend';
      const mockErrorResponseBody = {
        status: StatusCodes.BAD_REQUEST,
        code: ErrorCode.VALIDATION_ERROR,
        message: originalBackendErrorMessage,
        errors: [
          {
            field: 'openapiSpec.paths',
            message: 'Must contain at least one path',
            source: 'body',
          },
        ],
      };
      const mockFullErrorResponse = createErrorResponse(mockErrorResponseBody);

      const scope = nock(BASE_URL)
        .post('/api/deployments/from-openapi')
        .query(true) // Match any query string
        .reply(StatusCodes.BAD_REQUEST, mockFullErrorResponse);

      const source = OpenApiSource.fromObject(validOpenApiSpec);

      await expect(
        sdk.mcp.deploy({
          spec: source,
          name: 'Test API for Backend Validation Error',
        })
      ).rejects.toThrowErrorMatchingSnapshot();

      expect(scope.isDone()).toBe(true);
    });

    it('should throw SpecInvalidError for malformed request body', async () => {
      const malformedRequestErrorMessage = 'Validation error';
      const mockErrorResponseBody = {
        status: StatusCodes.BAD_REQUEST,
        code: ErrorCode.VALIDATION_ERROR,
        message: malformedRequestErrorMessage,
        errors: [
          {
            field: 'name',
            message: 'Expected string, received number',
            source: 'body',
          },
        ],
      };
      const mockFullErrorResponse = createErrorResponse(mockErrorResponseBody);

      const scope = nock(BASE_URL)
        .post('/api/deployments/from-openapi')
        .query(true)
        .reply(StatusCodes.BAD_REQUEST, mockFullErrorResponse);

      const source = OpenApiSource.fromObject(validOpenApiSpec);

      await expect(
        sdk.mcp.deploy({
          spec: source,
          name: 12345 as unknown as string, // Force invalid type for test
        })
      ).rejects.toThrowErrorMatchingSnapshot();

      expect(scope.isDone()).toBe(true);
    });

    it('should throw AuthError when using invalid API key', async () => {
      const mockErrorResponse = createErrorResponse({
        status: StatusCodes.UNAUTHORIZED,
        code: ErrorCode.AUTH_ERROR,
        message: 'Invalid API key from mock',
      });

      const scope = nock(BASE_URL)
        .post('/api/deployments/from-openapi')
        .query(true)
        .reply(StatusCodes.UNAUTHORIZED, mockErrorResponse);

      const source = OpenApiSource.fromObject(validOpenApiSpec);

      await expect(
        invalidSdk.mcp.deploy({
          spec: source,
          name: 'Auth Test Service (Nock)',
        })
      ).rejects.toThrowErrorMatchingSnapshot();

      expect(scope.isDone()).toBe(true);
    });

    it('should throw ApiError for server errors (e.g., 500)', async () => {
      const serverErrorMessage = 'Internal mock server error occurred';
      const mockErrorResponse = createErrorResponse({
        status: StatusCodes.INTERNAL_SERVER_ERROR,
        code: ErrorCode.INTERNAL_ERROR,
        message: serverErrorMessage,
      });

      const scope = nock(BASE_URL)
        .post('/api/deployments/from-openapi')
        .query(true)
        .reply(StatusCodes.INTERNAL_SERVER_ERROR, mockErrorResponse);

      const source = OpenApiSource.fromObject(validOpenApiSpec);

      await expect(
        sdk.mcp.deploy({
          spec: source,
          name: 'Server Error Test (Nock)',
        })
      ).rejects.toThrowErrorMatchingSnapshot();

      expect(scope.isDone()).toBe(true);
    });

    it('should throw ApiError for not found errors (404)', async () => {
      const notFoundMessage = 'Deployment not found';
      const mockErrorResponse = createErrorResponse({
        status: StatusCodes.NOT_FOUND,
        code: ErrorCode.NOT_FOUND,
        message: notFoundMessage,
        errors: [{ field: 'id', message: notFoundMessage, source: 'service' }],
      });

      const scope = nock(BASE_URL)
        .post('/api/deployments/from-openapi')
        .query(true)
        .reply(StatusCodes.NOT_FOUND, mockErrorResponse);

      const source = OpenApiSource.fromObject(validOpenApiSpec);

      await expect(
        sdk.mcp.deploy({
          spec: source,
          name: 'Not Found Test (Nock)',
        })
      ).rejects.toThrowErrorMatchingSnapshot();

      expect(scope.isDone()).toBe(true);
    });
  });
});
