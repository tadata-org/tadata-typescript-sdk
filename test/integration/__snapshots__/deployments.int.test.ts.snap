// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deployments Integration Test (Nock) when deploying an OpenAPI spec should throw ApiError for not found errors (404) 1`] = `"Deployment not found"`;

exports[`Deployments Integration Test (Nock) when deploying an OpenAPI spec should throw ApiError for server errors (e.g., 500) 1`] = `"Internal mock server error occurred"`;

exports[`Deployments Integration Test (Nock) when deploying an OpenAPI spec should throw AuthError when using invalid API key 1`] = `"Invalid API key from mock"`;

exports[`Deployments Integration Test (Nock) when deploying an OpenAPI spec should throw SpecInvalidError for backend validation failures 1`] = `"Invalid OpenAPI specification from backend"`;

exports[`Deployments Integration Test (Nock) when deploying an OpenAPI spec should throw SpecInvalidError for malformed request body 1`] = `"Validation error"`;
